CONTRIBUTING.md

```diff
@@ -0,0 +1,285 @@
+
+Some quick notes on testing terminology (still a WIP):
+
+ * Testing - anything done to test, whether manual, automated, or otherwise.
+ * Automated Testing - Any type of written test that can be run in an automated way, typically in the form of C# tests.
+ * Spec / Fact / Observation - these are synonyms for a test or validation.
+ * System Under Test (SUT) - the code or concern you are testing.
+ * Mock / Fake / Stub / Double - an object that provides a known state back to the system under test when the system under test interacts with other objects. This can be done with unit and whitebox integration testing. This allows for actual unit testing as most units (classes/functions) depend on working with other units (classes/functions) to get or set information and state. While each of [these are slightly different](https://martinfowler.com/articles/mocksArentStubs.html), the basic functionality is that they are standing in for other production code.
+ * Concern - an area of production code you are testing in e.g. "Concern for AutoUninstallerService".
+ * Regression Test Suite / Regression Suite - The automated tests that are in the form of code.
+ * Whitebox Testing - tests where you access the internals of the application.
+    * Unit Testing - We define a unit as a class and a method in C#. In PowerShell this is per function. If it involves another class or function, you have graduated to an integration. This is where Mocks come in to ensure no outside state is introduced.
+    * Whitebox Integration Testing - testing anything that is more than a unit.
+    * System Integration Testing - testing anything that goes out of the bounds of the written production code. Typically when running the code to get or set some state is where you will see this. And yes, even using DateTime.Now counts as system integration testing as it accesses something external to the application. This is why you will see we insulate those calls to something in the application so they can be easily tested against.
+ * Blackbox Testing - tests where you do not access internals of the application
+    * Physical Integration Testing - This is where you are testing the application with other components such as config files.
+    * Blackbox Integration Testing / End to End Testing - This is where you are testing inputs and outputs of the system.
+
+As far as testing goes, unit tests are extremely quick feedback and great for longer term maintenance, where black box tests give you the most coverage, but are the slowest feedback loops and typically the most frail. Each area of testing has strengths and weaknesses and it's good to understand each of them.
+
+**NOTE**: One of the hardest forms of automated testing is unit testing, as it almost always requires faking out other parts of the system (also known as mocking).
+
+#### Chocolatey Build

```

Comment:
Maybe the choco cli specific items could be factored out into a separate document (CONTRIB_PER_REPO.md), and linked to from here? And other repository-specific items in other repositories could be added.